agent/internal/fluent_intg_test.go

```diff
@@ -0,0 +1,257 @@
+
+	// THEN fluent should parse all fields as expected and ship them to elastic.
+	logs, _, err := elastic.TrialLogs(int(trialID), 0, 4, nil, nil)
+	assert.NilError(t, err, "failed to retrieve trial logs")
+	for i, l := range logs {
+		assertLogEquals(t, *l, expected[i])
+	}
+}
+
+func makeLogTestCase(trialID int, agentID string) ([]model.TrialLog, string) {
+	expected := []model.TrialLog{
+		{
+			TrialID:     trialID,
+			AgentID:     &agentID,
+			ContainerID: stringToPointer("goodcontainer"),
+			RankID:      intToPointer(4),
+			Log:         stringToPointer("\n"),
+			StdType:     stringToPointer("stdout"),
+		},
+		{
+			TrialID:     trialID,
+			AgentID:     &agentID,
+			ContainerID: stringToPointer("goodcontainer"),
+			RankID:      intToPointer(1),
+			Level:       stringToPointer("INFO"),
+			Log:         stringToPointer("Workload completed: <RUN_STEP (100 Batches): (580,6289,4)> (duration 0:00:01.496132)\n"), // nolint:lll // Anything to make these lines shorter would look worse.
+			StdType:     stringToPointer("stdout"),
+		},
+		{
+			TrialID:     trialID,
+			AgentID:     &agentID,
+			ContainerID: stringToPointer("goodcontainer"),
+			RankID:      intToPointer(2),
+			Level:       stringToPointer("ERROR"),
+			Log:         stringToPointer("Workload completed: <RUN_STEP (100 Batches): (580,6289,4)> (duration 9:99:99)\n"), // nolint:lll
+			StdType:     stringToPointer("stdout"),
+		},
+		{
+			TrialID:     trialID,
+			AgentID:     &agentID,
+			ContainerID: stringToPointer("goodcontainer"),
+			RankID:      intToPointer(3),
+			Log:         stringToPointer("urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f29a414dd30>: Failed to establish a new connection: [Errno 110] END\n"), // nolint:lll
+			StdType:     stringToPointer("stdout"),
+		},
+	}
+	actual := `[rank=4] 
+[rank=1] INFO: Workload completed: <RUN_STEP (100 Batches): (580,6289,4)> (duration 0:00:01.496132)
+[rank=2] ERROR: Workload completed: <RUN_STEP (100 Batches): (580,6289,4)> (duration 9:99:99)
+[rank=3] urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f29a414dd30>: Failed to establish a new connection: [Errno 110] END` // nolint:lll
+	return expected, actual
+}
+
+func mockLogAcceptor(t *testing.T) (chan model.TrialLog, func()) {
+	e := echo.New()
+	baseListener, err := net.Listen("tcp", fmt.Sprintf(":%d", 8080))
+	assert.NilError(t, err, "error starting mock master listener")
+	e.Listener = baseListener
+	logBuffer := make(chan model.TrialLog, 100)
+	e.POST("/trial_logs", func(ctx echo.Context) error {
+		body, err := ioutil.ReadAll(ctx.Request().Body)
+		if err != nil {
+			return err
+		}
+
+		var logs []model.TrialLog
+		if err = json.Unmarshal(body, &logs); err != nil {
+			return err
+		}
+
+		for _, l := range logs {
+			end := strings.Contains(*l.Log, "END")
+			logBuffer <- l
+			if end {
+				close(logBuffer)
```

Comment:
blocking: Can the `close` just happen after the loop once all the entries have been sent?

Reply:
I ended up doing something a little different. The close cant happen then because the channel persists between API hits and that would be a race with when the logs go and when when decides to flush.